homeassistant/components/media_player/songpal.py

```diff
@@ -0,0 +1,242 @@
+                  SUPPORT_VOLUME_MUTE | SUPPORT_SELECT_SOURCE | \
+                  SUPPORT_TURN_ON | SUPPORT_TURN_OFF
+
+_LOGGER = logging.getLogger(__name__)
+
+
+SET_SOUND_SETTING = "songpal_set_sound_setting"
+PARAM_NAME = "name"
+PARAM_VALUE = "value"
+
+CONF_ENDPOINT = "endpoint"
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_NAME, default=None): cv.string,
+    vol.Required(CONF_ENDPOINT): cv.string,
+})
+
+
+async def async_setup_platform(hass, config,
+                               async_add_devices, discovery_info=None):
+    """Set up the Songpal platform."""
+    devices = []
+    if discovery_info is not None:
+        _LOGGER.debug("Got autodiscovered device: %s", discovery_info)
+        devices.append(
+            SongpalDevice(discovery_info["name"],
+                          discovery_info["properties"]["endpoint"]))
+    else:
+        songpal = SongpalDevice(config.get(CONF_NAME),
+                                config.get(CONF_ENDPOINT))
+        devices.append(songpal)
+
+    async_add_devices(devices, True)
+
+    async def async_service_handler(service):
+        """Service handler."""
+        params = {key: value for key, value in service.data.items()
+                  if key != ATTR_ENTITY_ID}
+        entity_id = service.data.get("entity_id", None)
+        _LOGGER.debug("Calling %s (entity: %s) with params %s",
+                      service, entity_id, params)
+
+        for device in devices:
+            if entity_id is None or device.entity_id == entity_id:
+                await device.async_set_sound_setting(params[PARAM_NAME],
+                                                     params[PARAM_VALUE])
+
+    schema = vol.Schema({vol.Required(PARAM_NAME): cv.string,
+                         vol.Required(PARAM_VALUE): cv.string})
+    hass.services.async_register(
+        DOMAIN, SET_SOUND_SETTING, async_service_handler,
+        schema=schema)
+
+
+class SongpalDevice(MediaPlayerDevice):
+    """Class representing a Songpal device."""
+
+    def __init__(self, name, endpoint):
+        """Init."""
+        import songpal
+        self._name = name
+        self.endpoint = endpoint
+        self._dev = songpal.Protocol(self.endpoint)
+        self._sysinfo = None  # type: songpal.Sysinfo
+
+        self._state = False
+        self._available = False
+        self._initialized = False
+
+        self._volume_control = None
+        self._volume_min = 0
+        self._volume_max = 1
+        self._volume = 0
+        self._is_muted = False
+
+        self._sources = []
+
+    @property
+    def name(self):
+        """Return name of the device."""
+        return self._name
+
+    @property
+    def unique_id(self):
+        """Return an unique ID."""
+        return self._sysinfo.macAddr
+
+    @property
+    def available(self):
+        """Return availability of the device."""
+        return self._available
+
+    @property
+    def dev(self):
```

Comment:
Why? Why not just assign it to `self.dev` in the constructor?

Reply:
Indeed, changed that.